src/librustc_privacy/lib.rs

```diff
@@ -504,6 +504,291 @@ impl<'a, 'tcx> Visitor<'tcx> for NamePrivacyVisitor<'a, 'tcx> {
+                    let mut check = self.check_item(item.id);
+                    check.ty().generics().predicates();
+                    if trait_ref.is_some() {
+                        check.impl_trait_ref();
+                    }
+                }
+                for impl_item_ref in impl_item_refs {
+                    let impl_item = self.tcx.hir.impl_item(impl_item_ref.id);
+                    self.check_item(impl_item.id).generics().predicates().ty();
+                }
+            }
+        }
+
+        self.current_item = self.tcx.hir.local_def_id(item.id);
+        intravisit::walk_item(self, item);
+        self.current_item = orig_current_item;
+    }
+
+    fn visit_ty(&mut self, ty: &'tcx hir::Ty) {
+        let orig_current_item = self.current_item;
+
+        if let hir::TyImplTrait(..) = ty.node {
+            // Check the traits being exposed, as they're separate,
+            // e.g. `impl Iterator<Item=T>` has two predicates,
+            // `X: Iterator` and `<X as Iterator>::Item == T`,
+            // where `X` is the `impl Iterator<Item=T>` itself,
+            // stored in `predicates_of`, not in the `Ty` itself.
+            self.check_item(ty.id).predicates();
+        }
+
+        intravisit::walk_ty(self, ty);
+        self.current_item = orig_current_item;
+    }
+}
+
+impl<'a, 'tcx> TypeVisitor<'tcx> for TypePrivacyVisitor<'a, 'tcx> {
+    fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
+        let ty_def_id = match ty.sty {
+            ty::TyAdt(adt, ..) => Some(adt.did),
+            ty::TyFnDef(did, ..) => Some(did),
+            ty::TyDynamic(ref predicates, ..) => {
+                let is_private = predicates.skip_binder().iter().any(|predicate| {
+                    let def_id = match *predicate {
+                        ty::ExistentialPredicate::Trait(trait_ref) => trait_ref.def_id,
+                        ty::ExistentialPredicate::Projection(proj) => proj.trait_ref.def_id,
+                        ty::ExistentialPredicate::AutoTrait(def_id) => def_id,
+                    };
+                    !self.item_is_accessible(def_id)
+                });
+                if is_private {
+                    let msg = format!("type `{}` is private", ty);
+                    self.tcx.sess.span_err(self.span, &msg);
+                    return true;
+                }
+                None
+            }
+            ty::TyAnon(..) => {
+                // FIXME: How to visit `impl Trait` type?
+                None
+            }
+            _ => None
+        };
+
+        if let Some(def_id) = ty_def_id {
+            if !self.item_is_accessible(def_id) {
+                let msg = format!("type `{}` is private", ty);
+                self.tcx.sess.span_err(self.span, &msg);
+                return true;
+            } else if let ty::TyFnDef(..) = ty.sty {
+                // Inherent static methods don't have self type in substs,
+                // we have to check it additionally.
```

Comment:
It's not because `Self` is a mere alias in them. The substitutions for impls (inherent *or* trait impls, but you never get the latter from type-checking, only e.g. trans or MIR inlining can see them) are *literally* for the `<...>` in `impl<...>`.

*However*, it should be impossible to access such an inherent method without one of:
* method call, which means there is an expression containing the `Self` type
* `T::method` which is sugar for `<T>::method`, *and* the semantic version of `T` *can be read* (see `rustdoc` and `rustc_save_analysis`)